### ISSUE TITLE
Mask loss in ssd_meta_arch

### ISSUE THREAD

**Issue by anshkumar** [tensorflow/models#8552]
I'm adding mask head to the ssd meta architecture. For this I've added necessary changes in the protos and convolution head. Now, the `prediction_dict` has a `mask_predictions` in it using [ConvolutionalMaskHead](https://github.com/tensorflow/models/blob/master/research/object_detection/predictors/heads/mask_head.py#L186) with little custom changes. SSD_meta_arch [postprocess](https://github.com/tensorflow/models/blob/master/research/object_detection/meta_architectures/ssd_meta_arch.py#L687) seems to handle mask head. But when looking at the [loss](https://github.com/tensorflow/models/blob/master/research/object_detection/meta_architectures/ssd_meta_arch.py#L839) function, it seems that it's not handling the mask loss. 

In the mask-rcnn the [_loss_box_classifier](https://github.com/tensorflow/models/blob/master/research/object_detection/meta_architectures/faster_rcnn_meta_arch.py#L2336) is defining the mask loss in the following way:

```
      second_stage_mask_loss = None
      if prediction_masks is not None:
        if groundtruth_masks_list is None:
          raise ValueError('Groundtruth instance masks not provided. '
                           'Please configure input reader.')

        if not self._is_training:
          (proposal_boxes, proposal_boxlists, paddings_indicator,
           one_hot_flat_cls_targets_with_background
          ) = self._get_mask_proposal_boxes_and_classes(
              detection_boxes, num_detections, image_shape,
              groundtruth_boxlists, groundtruth_classes_with_background_list,
              groundtruth_weights_list)
        unmatched_mask_label = tf.zeros(image_shape[1:3], dtype=tf.float32)
        (batch_mask_targets, _, _, batch_mask_target_weights,
         _) = target_assigner.batch_assign_targets(
             target_assigner=self._detector_target_assigner,
             anchors_batch=proposal_boxlists,
             gt_box_batch=groundtruth_boxlists,
             gt_class_targets_batch=groundtruth_masks_list,
             unmatched_class_label=unmatched_mask_label,
             gt_weights_batch=groundtruth_weights_list)

        # Pad the prediction_masks with to add zeros for background class to be
        # consistent with class predictions.
        if prediction_masks.get_shape().as_list()[1] == 1:
          # Class agnostic masks or masks for one-class prediction. Logic for
          # both cases is the same since background predictions are ignored
          # through the batch_mask_target_weights.
          prediction_masks_masked_by_class_targets = prediction_masks
        else:
          prediction_masks_with_background = tf.pad(
              prediction_masks, [[0, 0], [1, 0], [0, 0], [0, 0]])
          prediction_masks_masked_by_class_targets = tf.boolean_mask(
              prediction_masks_with_background,
              tf.greater(one_hot_flat_cls_targets_with_background, 0))

        mask_height = shape_utils.get_dim_as_int(prediction_masks.shape[2])
        mask_width = shape_utils.get_dim_as_int(prediction_masks.shape[3])
        reshaped_prediction_masks = tf.reshape(
            prediction_masks_masked_by_class_targets,
            [batch_size, -1, mask_height * mask_width])

        batch_mask_targets_shape = tf.shape(batch_mask_targets)
        flat_gt_masks = tf.reshape(batch_mask_targets,
                                   [-1, batch_mask_targets_shape[2],
                                    batch_mask_targets_shape[3]])

        # Use normalized proposals to crop mask targets from image masks.
        flat_normalized_proposals = box_list_ops.to_normalized_coordinates(
            box_list.BoxList(tf.reshape(proposal_boxes, [-1, 4])),
            image_shape[1], image_shape[2], check_range=False).get()

        flat_cropped_gt_mask = self._crop_and_resize_fn(
            tf.expand_dims(flat_gt_masks, -1),
            tf.expand_dims(flat_normalized_proposals, axis=1),
            [mask_height, mask_width])
        # Without stopping gradients into cropped groundtruth masks the
        # performance with 100-padded groundtruth masks when batch size > 1 is
        # about 4% worse.
        # TODO(rathodv): Investigate this since we don't expect any variables
        # upstream of flat_cropped_gt_mask.
        flat_cropped_gt_mask = tf.stop_gradient(flat_cropped_gt_mask)

        batch_cropped_gt_mask = tf.reshape(
            flat_cropped_gt_mask,
            [batch_size, -1, mask_height * mask_width])

        mask_losses_weights = (
            batch_mask_target_weights * tf.cast(paddings_indicator,
                                                dtype=tf.float32))
        mask_losses = self._second_stage_mask_loss(
            reshaped_prediction_masks,
            batch_cropped_gt_mask,
            weights=tf.expand_dims(mask_losses_weights, axis=-1),
            losses_mask=losses_mask)
        total_mask_loss = tf.reduce_sum(mask_losses)
        normalizer = tf.maximum(
            tf.reduce_sum(mask_losses_weights * mask_height * mask_width), 1.0)
        second_stage_mask_loss = total_mask_loss / normalizer

      if second_stage_mask_loss is not None:
        mask_loss = tf.multiply(self._second_stage_mask_loss_weight,
                                second_stage_mask_loss, name='mask_loss')
        loss_dict[mask_loss.op.name] = mask_loss
```

The tensor of SSD is very much different from that of faster rcnn architecture. I need help with the implementation of mask loss in SSD.

**Comment by anshkumar** [comment]
Okay, I've added mask loss in `ssd_meta_arch.py`. The changes looks like this:

```
if 'mask_predictions' in prediction_dict:
  prediction_masks = prediction_dict['mask_predictions']
  if not self.groundtruth_has_field(fields.BoxListFields.masks):
    raise ValueError('Groundtruth instance masks not provided. '
                     'Please configure input reader.')
  groundtruth_boxlists = [
      box_list.BoxList(boxes) for boxes in \
      self.groundtruth_lists(fields.BoxListFields.boxes)
  ]

  unmatched_mask_label = tf.zeros(true_image_shapes[0,:2], 
    dtype=tf.float32)

  (batch_mask_targets, _, _, batch_mask_target_weights,
     _) = target_assigner.batch_assign_targets(
      self._target_assigner,
      self.anchors,
      groundtruth_boxlists,
      self.groundtruth_lists(fields.BoxListFields.masks),
      unmatched_mask_label,
      weights)

  mask_height = shape_utils.get_dim_as_int(prediction_masks.shape[3])
  mask_width = shape_utils.get_dim_as_int(prediction_masks.shape[4])
  batch_size = shape_utils.get_dim_as_int(prediction_masks.shape[0])

  reshaped_prediction_masks = tf.reshape(
      prediction_masks,
      [batch_size, -1, mask_height * mask_width])

  batch_mask_targets_shape = tf.shape(batch_mask_targets)

  flat_gt_masks = tf.reshape(batch_mask_targets,
                               [-1, batch_mask_targets_shape[2],
                                batch_mask_targets_shape[3]])

  image_shape = true_image_shapes[0,:2]

  flat_normalized_proposals = box_list_ops.to_normalized_coordinates(
        box_list.BoxList(
          tf.reshape(prediction_dict['box_encodings'], [-1, 4])),
        image_shape[0], image_shape[1], check_range=False).get()

  flat_cropped_gt_mask = self._crop_and_resize_fn(
        tf.expand_dims(flat_gt_masks, -1),
        tf.expand_dims(flat_normalized_proposals, axis=1),
        [mask_height, mask_width])

  flat_cropped_gt_mask = tf.stop_gradient(flat_cropped_gt_mask)

  batch_cropped_gt_mask = tf.reshape(
        flat_cropped_gt_mask,
        [batch_size, -1, mask_height * mask_width])

  mask_losses = self._mask_loss(
        reshaped_prediction_masks,
        batch_cropped_gt_mask,
        weights=tf.expand_dims(batch_mask_target_weights, axis=-1),
        losses_mask=losses_mask)
  total_mask_loss = tf.reduce_sum(mask_losses)

  mask_loss = tf.multiply((self._mask_loss_weight /
                                   normalizer),
                                  total_mask_loss,
                                  name='mask_loss')

  loss_dict['Loss/mask_loss'] = mask_loss
```
But on running the training I'm getting following error:

```
2020-06-02 13:26:05.325204: F ./tensorflow/core/util/gpu_launch_config.h:129] Check failed: work_element_count > 0 (-1298327296 vs. 0)
Fatal Python error: Aborted

Thread 0x00007f779cff9700 (most recent call first):
  File "/usr/lib/python3.6/threading.py", line 295 in wait
  File "/usr/lib/python3.6/queue.py", line 164 in get
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/summary/writer/event_file_writer.py", line 159 in run
  File "/usr/lib/python3.6/threading.py", line 916 in _bootstrap_inner
  File "/usr/lib/python3.6/threading.py", line 884 in _bootstrap

Thread 0x00007f7835c71740 (most recent call first):
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/client/session.py", line 1443 in _call_tf_sessionrun
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/client/session.py", line 1350 in _run_fn
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/client/session.py", line 1365 in _do_call
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/client/session.py", line 1359 in _do_run
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/client/session.py", line 1180 in _run
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/client/session.py", line 956 in run
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/training/monitored_session.py", line 1176 in run
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/training/monitored_session.py", line 1418 in run
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/training/monitored_session.py", line 1345 in run
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/training/monitored_session.py", line 1259 in run
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/training/monitored_session.py", line 754 in run
  File "/home/sort/.local/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 1494 in _train_with_estimator_spec
  File "/home/sort/.local/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 1195 in _train_model_default
  File "/home/sort/.local/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 1161 in _train_model
  File "/home/sort/.local/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 370 in train
  File "/home/sort/.local/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/training.py", line 714 in run_local
  File "/home/sort/.local/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/training.py", line 613 in run
  File "/home/sort/.local/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/training.py", line 473 in train_and_evaluate
  File "object_detection/model_main.py", line 134 in main
  File "/home/sort/.local/lib/python3.6/site-packages/absl/app.py", line 250 in _run_main
  File "/home/sort/.local/lib/python3.6/site-packages/absl/app.py", line 299 in run
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/platform/app.py", line 40 in run
  File "object_detection/model_main.py", line 138 in <module>
./train_ssd.sh: line 4:  4967 Aborted                 (core dumped) CUDA_VISIBLE_DEVICES=0 python3 object_detection/model_main.py --logtostderr --model_dir=/home/sort/ved/research/ssd_mobilenet_v2_coco --pipeline_config_path=/home/sort/ved/research/ssd_mobilenet_v2_coco.config
```

I don't really know what the error means. Any help ?
The complete code of it can be found here ([object detection](https://github.com/anshkumar/object_detection), [slim](https://github.com/anshkumar/slim))

Here is my configurations: 

Tensorflow version: 1.15.2
NVIDIA Driver Version: 440.64.00
CUDA Version: 10.2
GPU: Tesla V100 PCIe 32GB

**Comment by guo654** [comment]
maybe the reason of your error is here>https://github.com/matterport/Mask_RCNN/issues/521
